File: homing-demo/src/main/java/hue/captains/singapura/js/homing/demo/playground/AnimalsSecretaryModule.java
package hue.captains.singapura.js.homing.demo.playground;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;

import java.util.List;

/**
 * The Secretary for the {@code AnimalsParty} — the {@code AnimalsPlayground}
 * workspace's first downstream ActorParty. Extracted as a standalone JS
 * module per the Diligent Secretaries doctrine's third pillar: pure
 * behaviour functions must be importable in isolation so a GraalVM-backed
 * JUnit harness can exercise them without a DOM.
 *
 * <p>Exports a single {@code AnimalsSecretary} JS object with two members:
 * {@code initial} (starting state record) and {@code behavior} (the pure
 * {@code (state, envelope) → Step} routing function). The
 * {@link AnimalsPlaygroundChrome} imports both and passes them to the
 * {@code Party} constructor.</p>
 *
 * <h2>State shape (Diligent Secretaries enrichment)</h2>
 *
 * <pre>{@code
 * {
 *     selectedAnimal: String,             // "-1" (Random) | concrete animal id
 *     lastChangedBy : AgentId | null,     // provenance of most recent change
 *     recentUnknown : [{ kind, from }]    // last N unrecognised messages (bounded at 10)
 * }
 * }</pre>
 *
 * <h2>Message kinds</h2>
 *
 * <ul>
 *   <li>{@code AnimalSelectionRequested(animal)} — record selection, broadcast
 *       {@code AnimalChanged(animal)} to all members.</li>
 *   <li>{@code CurrentAnimalRequested} — late-join sync. If
 *       {@code selectedAnimal} is non-default, reply to the asking member via
 *       {@code SendToMember(AnimalChanged)}; if still {@code '-1'}, no-op
 *       (asking widget keeps its own per-instance default).</li>
 *   <li>Anything else — append to {@code recentUnknown} (bounded at 10);
 *       no action emitted.</li>
 * </ul>
 *
 * @since RFC 0028 cycle 6 phase 1 — Secretary-as-module refactor for JS test harness
 */
public record AnimalsSecretaryModule() implements DomModule<AnimalsSecretaryModule> {

    /** The single export — a JS object with {@code initial} and {@code behavior} members. */
    public record AnimalsSecretary() implements Exportable._Constant<AnimalsSecretaryModule> {}

    public static final AnimalsSecretaryModule INSTANCE = new AnimalsSecretaryModule();

    @Override
    public ImportsFor<AnimalsSecretaryModule> imports() {
        return ImportsFor.<AnimalsSecretaryModule>builder().build();
    }

    @Override
    public ExportsOf<AnimalsSecretaryModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new AnimalsSecretary()));
    }
}

File: homing-demo/src/main/resources/homing/js/hue/captains/singapura/js/homing/demo/playground/AnimalsSecretaryModule.js
// =============================================================================
// AnimalsSecretaryModule — the Secretary for AnimalsPlayground's AnimalsParty.
//
// Per the Diligent Secretaries doctrine: pure (state, envelope) -> Step
// behavior plus initial state, both exposed as members of a single
// top-level object for clean framework import + isolated test harness load.
//
// Diligence enrichment in state:
//   - selectedAnimal : current value ('-1' = Random default, else animal id)
//   - lastChangedBy  : provenance of the most recent state change (envelope.from)
//   - recentUnknown  : last 10 unrecognised messages (kind + origin), bounded
//
// Side-effect-free; no DOM, no console, no captures. Per the
// Lambdas-Are-Functional-Object-Shortcuts doctrine, this entire module
// is a Functional Object simplification — pure behavior + immutable state shape.
// =============================================================================

var AnimalsSecretary = {

    initial: {
        selectedAnimal: "-1",
        lastChangedBy : null,
        recentUnknown : []
    },

    /**
     * Pure routing function. (state, envelope) -> { newState, actions }.
     * Same inputs always produce same outputs; no side effects.
     */
    behavior: function (state, envelope) {
        var msg = envelope.message;

        switch (msg.kind) {

            case "AnimalSelectionRequested": {
                return {
                    newState: {
                        selectedAnimal: msg.animal,
                        lastChangedBy : envelope.from,
                        recentUnknown : state.recentUnknown
                    },
                    actions: [{
                        kind   : "BroadcastToMembers",
                        message: { kind: "AnimalChanged", animal: msg.animal }
                    }]
                };
            }

            case "GameActionRequested": {
                // RFC 0028 broadcast feature — a live game event from the
                // MovingAnimalPlayWidget (or a directed snapshot reply). When
                // `to` is present, deliver privately to that one member (the
                // late-join snapshot path); otherwise broadcast to all members.
                // The play source has no GameAction reactor, so it harmlessly
                // ignores the echo of its own broadcast. State is untouched —
                // the Secretary routes game traffic, it does not store it
                // (history would be unbounded; the snapshot path covers
                // late-join instead).
                if (msg.to) {
                    return {
                        newState: state,
                        actions: [{
                            kind   : "SendToMember",
                            to     : msg.to,
                            message: { kind: "GameAction", event: msg.event }
                        }]
                    };
                }
                return {
                    newState: state,
                    actions: [{
                        kind   : "BroadcastToMembers",
                        message: { kind: "GameAction", event: msg.event }
                    }]
                };
            }

            case "CurrentGameRequested": {
                // Late-join sync for replay widgets. Fan the request out to
                // members; the live play widget answers with a directed
                // GameActionRequested(to: asker, event: Snapshot). Replays and
                // the asker itself have no GameSnapshotRequested reactor and
                // ignore it. Carries the asker's id so the play side can reply
                // privately.
                return {
                    newState: state,
                    actions: [{
                        kind   : "BroadcastToMembers",
                        message: { kind: "GameSnapshotRequested", asker: envelope.from }
                    }]
                };
            }

            case "CurrentAnimalRequested": {
                // Late-join sync. If a concrete animal has been chosen, reply
                // privately to the asker reusing the AnimalChanged fact shape —
                // the widget's existing reactor handles both paths uniformly.
                // If still at the '-1' Random default, no useful answer; the
                // widget keeps its own per-instance default.
                if (state.selectedAnimal !== "-1") {
                    return {
                        newState: state,
                        actions: [{
                            kind   : "SendToMember",
                            to     : envelope.from,
                            message: { kind: "AnimalChanged", animal: state.selectedAnimal }
                        }]
                    };
                }
                return { newState: state, actions: [] };
            }

            default: {
                // Diligent defensive observability: track unknowns in state so
                // they surface via inspect() rather than silently disappearing.
                // Bounded at 10 entries via .slice(-10) so state cannot grow.
                var recent = state.recentUnknown.concat([{
                    kind: msg.kind,
                    from: envelope.from
                }]).slice(-10);
                return {
                    newState: {
                        selectedAnimal: state.selectedAnimal,
                        lastChangedBy : state.lastChangedBy,
                        recentUnknown : recent
                    },
                    actions: []
                };
            }
        }
    }
};
